job-service-scheduled-executor/src/main/java/com/github/jobservice/scheduledexecutor/QueueServicesFactory.java

```diff
@@ -47,42 +59,100 @@ public final class QueueServicesFactory
         }
         QUEUE_ARGUMENTS.put(RABBIT_PROP_QUEUE_TYPE, queueType);
+
+        // Register shutdown hook to clean up connection on JVM shutdown
+        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
+            final Connection connection = SHARED_CONNECTION.get();
+            if (connection != null && connection.isOpen()) {
+                try {
+                    connection.close(5_000);
+                } catch (final IOException e) {
+                }
+            }
+        }));
     }
 
     /**
-     * Create a new QueueServices object.
+     * Create a new QueueServices object using the shared connection.
      *
-     * @param   targetQueue                     the target queue
      * @param   partitionId                     the partition ID
+     * @param   jobId                           the job ID
+     * @param   targetQueue                     the target queue
      * @param   codec                           the serialization codec
-     * @return  ScheduledExecutorQueueServices  new ScheduledExecutorQueueServices object
-     * @throws  IOException                     thrown if the connection cannot be created
+     * @return  QueueServices                   new QueueServices object
+     * @throws  IOException                     thrown if the connection/channel cannot be created
      * @throws  TimeoutException                thrown if the connection cannot be created
      */
-    public static QueueServices create(final String targetQueue, final String partitionId, final Codec codec)
+    public static QueueServices create(
+            final String partitionId, final String jobId, final String targetQueue, final Codec codec)
             throws IOException, TimeoutException, URISyntaxException, NoSuchAlgorithmException, KeyManagementException
     {
-        //  Create connection and channel for publishing messages.
-        LOG.debug("Creating connection ...");
-        final Connection connection = createConnection();
+        // Get or create the shared connection
+        final Connection connection = getOrCreateSharedConnection();
 
-        LOG.debug("Creating channel ...");
+        // Create a new channel for this job (channels are not thread-safe, so each job needs its own)
+        LOG.debug("Creating channel for job [partitionId={}, jobId={}, queue={}]...",
+                partitionId, jobId, targetQueue);
         final Channel publishChannel = connection.createChannel();
+        if (publishChannel == null) {
+            throw new IOException(
+                    MessageFormat.format(
+                            "Failed to create a new channel for job [partitionId={0}, jobId={1}, queue={2}]. " +
+                            "Job will be retried.", partitionId, jobId, targetQueue));
+        }
+
+        publishChannel.confirmSelect();
+        RabbitMqAsyncListener.registerListeners(publishChannel, partitionId, jobId, targetQueue);
```

Comment:
Doesn't the `.confirmSelect()` enable synchronous acknowledgments, defeating the purpose of the async-listening on the next line?

Reply:
Looking at this a bit more you also require `.waitForConfirmsOrDie()` for synchronous listening, like we do in [`job-service-core/.../QueueServices.java:148`](https://github.com/JobService/job-service/blob/v10.0.0/job-service-core/src/main/java/com/github/jobservice/core/queue/QueueServices.java#L148).  Is that something we could do here or do we require the complexity of asynchronous listening for performance reasons?